File: packages/sdk-typescript/src/crypto.ts
/**
 * Cryptographic utilities for URL signing and encryption.
 *
 * Uses Node.js built-in `crypto` module — zero external dependencies.
 * These are pure functions with no network calls, so link generation
 * is instantaneous and works offline.
 */

import { createHmac, createCipheriv, createDecipheriv, randomBytes } from "node:crypto";

/**
 * Generate an HMAC-SHA256 signature for a URL string.
 *
 * @param secret - The shared HMAC secret (must match server's `HMAC_SECRET`).
 * @param url - The destination URL to sign.
 * @returns Lowercase hex-encoded signature string.
 */
export function signHmac(secret: string, url: string): string {
  return createHmac("sha256", secret).update(url).digest("hex");
}

/**
 * Verify an HMAC-SHA256 signature against a URL string.
 *
 * @param secret - The shared HMAC secret.
 * @param url - The destination URL that was signed.
 * @param signature - The hex-encoded signature to verify.
 * @returns `true` if the signature is valid.
 */
export function verifyHmac(secret: string, url: string, signature: string): boolean {
  const expected = signHmac(secret, url);
  // Constant-time comparison to prevent timing attacks
  if (expected.length !== signature.length) return false;
  let result = 0;
  for (let i = 0; i < expected.length; i++) {
    result |= expected.charCodeAt(i) ^ signature.charCodeAt(i);
  }
  return result === 0;
}

/**
 * Encrypt a URL using AES-256-GCM.
 *
 * Returns a base64url-encoded string containing `nonce || ciphertext || tag`.
 * Compatible with the Rust core's `decrypt_url()`.
 *
 * @param keyHex - 64-character hex string (32 bytes) for the AES key.
 * @param url - The destination URL to encrypt.
 * @returns Base64url-encoded encrypted blob (no padding).
 */
export function encryptUrl(keyHex: string, url: string): string {
  const key = Buffer.from(keyHex, "hex");
  if (key.length !== 32) {
    throw new Error(`Encryption key must be 32 bytes (got ${key.length})`);
  }

  const nonce = randomBytes(12); // 96-bit nonce for GCM
  const cipher = createCipheriv("aes-256-gcm", key, nonce);

  const encrypted = Buffer.concat([cipher.update(url, "utf8"), cipher.final()]);
  const tag = cipher.getAuthTag(); // 16 bytes

  // Format: nonce (12) || ciphertext || tag (16)
  // This matches the Rust core's format: nonce || ciphertext+tag
  // (aes-gcm crate appends tag to ciphertext automatically)
  const combined = Buffer.concat([nonce, encrypted, tag]);

  // Base64url encoding without padding (matches Rust's URL_SAFE_NO_PAD)
  return combined.toString("base64url");
}

/**
 * Decrypt a base64url-encoded AES-256-GCM payload back to the original URL.
 *
 * @param keyHex - 64-character hex string (32 bytes) for the AES key.
 * @param encoded - Base64url-encoded encrypted blob from `encryptUrl()`.
 * @returns The original plaintext URL.
 */
export function decryptUrl(keyHex: string, encoded: string): string {
  const key = Buffer.from(keyHex, "hex");
  const combined = Buffer.from(encoded, "base64url");

  if (combined.length < 28) {
    // 12 (nonce) + 16 (tag) minimum
    throw new Error("Ciphertext too short");
  }

  const nonce = combined.subarray(0, 12);
  const ciphertext = combined.subarray(12, combined.length - 16);
  const tag = combined.subarray(combined.length - 16);

  const decipher = createDecipheriv("aes-256-gcm", key, nonce);
  decipher.setAuthTag(tag);

  const decrypted = Buffer.concat([decipher.update(ciphertext), decipher.final()]);
  return decrypted.toString("utf8");
}

File: src/crypto.rs
//! Cryptographic utilities for URL signing and encryption.
//!
//! Provides two security modes for the `/t` click-tracking endpoint:
//!
//! - **Signed (HMAC-SHA256):** The destination URL is visible in the query string
//!   and protected by a hex-encoded HMAC signature. Fast to verify, transparent.
//! - **Encrypted (AES-256-GCM):** The destination URL is encrypted into an opaque
//!   base64url blob. Provides confidentiality — the URL is invisible to the end user.
//!
//! The SDK uses [`sign_hmac()`] or [`encrypt_url()`] to generate links;
//! the core uses [`verify_hmac()`] or [`decrypt_url()`] to validate them.

use hmac::{Hmac, Mac};
use sha2::Sha256;

/// HMAC-SHA256 type alias for convenience.
type HmacSha256 = Hmac<Sha256>;

/// Verify an HMAC-SHA256 signature against a URL string.
///
/// Returns `true` if the hex-encoded `signature` matches the HMAC of `url`
/// computed with `secret`. Returns `false` on any error (bad hex, wrong key, etc.).
pub fn verify_hmac(secret: &str, url: &str, signature: &str) -> bool {
    let Ok(mut mac) = <HmacSha256 as Mac>::new_from_slice(secret.as_bytes()) else {
        return false;
    };
    mac.update(url.as_bytes());
    let Ok(sig_bytes) = hex::decode(signature) else {
        return false;
    };
    mac.verify_slice(&sig_bytes).is_ok()
}

/// Generate an HMAC-SHA256 signature for a URL string.
///
/// Returns the signature as a lowercase hex string. Used by the SDK
/// to sign tracking URLs before distribution.
pub fn sign_hmac(secret: &str, url: &str) -> String {
    let mut mac = <HmacSha256 as Mac>::new_from_slice(secret.as_bytes())
        .expect("HMAC can take key of any size");
    mac.update(url.as_bytes());
    hex::encode(mac.finalize().into_bytes())
}

// --- AES-256-GCM encryption/decryption for opaque URL mode ---

use aes_gcm::{
    aead::{Aead, KeyInit, OsRng},
    Aes256Gcm, AeadCore, Nonce,
};
use base64::{engine::general_purpose::URL_SAFE_NO_PAD, Engine};

/// Encrypt a URL using AES-256-GCM.
///
/// Generates a random 12-byte nonce, encrypts the URL, and returns a
/// base64url-encoded string containing `nonce || ciphertext || tag`.
/// Used by the SDK to create opaque tracking links.
pub fn encrypt_url(key: &[u8], url: &str) -> Result<String, String> {
    let cipher = Aes256Gcm::new_from_slice(key).map_err(|e| format!("Invalid key: {e}"))?;
    let nonce = Aes256Gcm::generate_nonce(&mut OsRng);
    let ciphertext = cipher
        .encrypt(&nonce, url.as_bytes())
        .map_err(|e| format!("Encryption failed: {e}"))?;

    // Prepend nonce (12 bytes) to ciphertext
    let mut combined = nonce.to_vec();
    combined.extend_from_slice(&ciphertext);
    Ok(URL_SAFE_NO_PAD.encode(&combined))
}

/// Decrypt a base64url-encoded AES-256-GCM payload back to the original URL.
///
/// Splits the first 12 bytes as the nonce, decrypts the remainder, and
/// returns the plaintext URL. Returns an error if the key is wrong,
/// the data is tampered, or the payload is malformed.
pub fn decrypt_url(key: &[u8], encoded: &str) -> Result<String, String> {
    let combined = URL_SAFE_NO_PAD
        .decode(encoded)
        .map_err(|e| format!("Invalid base64: {e}"))?;

    if combined.len() < 12 {
        return Err("Ciphertext too short".into());
    }

    let (nonce_bytes, ciphertext) = combined.split_at(12);
    let nonce = Nonce::from_slice(nonce_bytes);
    let cipher = Aes256Gcm::new_from_slice(key).map_err(|e| format!("Invalid key: {e}"))?;
    let plaintext = cipher
        .decrypt(nonce, ciphertext)
        .map_err(|_| "Decryption failed: invalid key or tampered data".to_string())?;

    String::from_utf8(plaintext).map_err(|e| format!("Invalid UTF-8: {e}"))
}
